### ISSUE TITLE
IndexOutOfRangeException thrown in VtNetCore

### ISSUE THREAD

**Issue by shimat** [GoogleCloudPlatform/iap-desktop#545]
When I was using IAP Desktop to connect to a GCE instance via SSH, the following error occurred.

```
System.IndexOutOfRangeException: Buffer does not contain enough data to process request
   場所 VtNetCore.XTermParser.XTermInputBuffer.PeekAhead(Int32 skip)
   場所 VtNetCore.XTermParser.DataConsumer.Push(Byte[] data)
   場所 Google.Solutions.IapDesktop.Extensions.Shell.Controls.VirtualTerminal.ReceiveData(String text)
   場所 Google.Solutions.IapDesktop.Extensions.Shell.Views.SshTerminal.SshTerminalPane.OnDataReceivedFromServerAsync(Object sender, DataReceivedEventArgs args)
   場所 Google.Solutions.IapDesktop.Extensions.Shell.Views.SshTerminal.SshTerminalPaneViewModel.<>c__DisplayClass38_1.<ConnectAsync>b__6()
```

https://github.com/darrenstarr/VtNetCore/blob/060a72f074aafb8f8720f41616727a69755dade7/VtNetCore/XTermParser/XTermInputBuffer.cs#L133

IAP Desktop Version 2.17.616.0
.NET 4.8.4300.0

**Comment by jpassing** [comment]
Did this exception cause any error in IAP Desktop? Or did you merely observe this exception in the debugger?

vtnetcore internally throws an IndexOutOfRangeException whenever the current buffer contains an incomplete xterm sequence. The library then catches the exception, reads some more data, and tries again. So seeing some IndexOutOfRangeException fly by is normal.

**Comment by shimat** [comment]
> Did this exception cause any error in IAP Desktop? 

Yes. An error dialog was displayed. I forgot to save its screenshot, but it looks like this: https://answers.microsoft.com/en-us/windows/forum/all/microsoft-net-framework-error/9533e7d4-5b76-47fd-a057-4950af9c3a8c

After pressing "Continue", I was able to continue using the application without crashing. So this problem is not serious, but in my environment, it happens once every few hours.

**Comment by jpassing** [comment]
After double-checking the code, I think this might be caused by an UTF-8 sequence that vtnetcore considers invalid.

Is the error reproducible? If yes, it would be very helpful if you could capture and share some of the received data:

1. Enable logging (Tools > Enable logging)
2. Reproduce the error
3. Hold the **Shift** key while right-clicking the tab of the SSH session
4. Select **Copy received data**
5. Disable logging (Tools > Enable logging)

The clipboard now contains a capture of the raw data received in your SSH session. The culprit should be close to the end of the capture. If the data isn't confidential, it would be great if you could share the last few lines.

**Comment by jpassing** [comment]
Nevermind, I'm able to reproduce the issue now.

**Comment by jpassing** [comment]
Certain Unicode characters such as `U+201B SINGLE HIGH-REVERSED-9 QUOTATION MARK` caused vtnetcore to wrongly interpret the following characters as an xterm control sequence (because `1b` is the Escape character), which in turn could lead to an IndexOutOfRangeException. I'm pretty confident that the errors you've seen have been caused by this issue.

I implemented a fix and it will go into the next release.
